File: algorithms/grokking-algorithms/intro.py
import numpy as np
import math

def get_halves(num):
    sum = num
    i = 1
    print(f"{0} {num} start")
    while math.ceil(sum) != 1:
        sum /= 2
        print(f"{i} {math.ceil(sum)}")
        i += 1


def binary_search(arr, item):
    low = 0
    high = len(arr) - 1
    print(f"searching for {item}, in a list of: {arr}")
    while low <= high:
        mid = (low + high) // 2
        guess = arr[mid]
        print(f"low: {low} high: {high} mid: {mid}, guess: {guess}")
        if guess == item:
            return mid
        elif guess > item:
            high = mid - 1
        else:
            low = mid + 1
    return None

def binary_search_DC2(arr, item):
    low = 0
    high = len(arr) - 1
    mid = (low + high) // 2
    guess = arr[mid]
    print(f"arr: {arr}, guess: {guess}")
    if item == guess:
        print("found it!")
        return mid
    else:
        print(f"continue searching in {arr[mid+1:] if item > guess else arr[0:mid-1]}...")
        return binary_search_DC2(arr[mid+1:], item) if item > guess else binary_search_DC2(arr[0:mid], item)

    return None

def binary_search_DC(arr, item):
    print("----------")
    print(f"starting search for {item}, in a list of: {arr}")
    if len(arr) == 2:
        print(f"found it!")
        return 0 if arr[0] == item else 1
    else:
        guess, low = 0, 0
        high_total = len(arr) - 1
        high_half = (low + high_total) // 2
        tmp_arr = arr[:high_half+1]
        print(f"searching for {item}, in the lower half: {tmp_arr}")
        while low <= high_half:
            mid = (low + high_half) // 2
            guess = tmp_arr[mid]
            print(f"low: {low} high: {high_half} mid: {mid}, guess: {guess}")
            if guess == item:
                return mid
            elif guess > item:
                high_half = mid - 1
            else:
                low = mid + 1
        if guess == item:
            return guess
        else:
           print("searching upper half..")
           return (high_half+1) + binary_search_DC(arr[high_half+1:], item)

    return None

def main():
    list = [1, 3, 5, 7, 9, 11, 2, 124, 43, 34]


    # print(list[3:])
    #print("binary_search:")
    #print(binary_search(sorted(list), 124))
    #print("=========================")
    #print("binary_search_DC (divide & conquer):")
    #print(binary_search_DC(sorted(list), 124))
    print("binary_search_DC2 (divide & conquer):")
    print(binary_search_DC2(sorted(list), 124))
    # # get_halves(28)


if __name__ == "__main__":
    main()

File: algorithms/grokking-algorithms/intro.c
#include <stdio.h>
#include <math.h>

static int binarySearch(int *arr, size_t arrLen, int elem){
  int low = 0;
  int high = arrLen - 1;
  printf("----------------\n");
  while(low <= high){
    int mid = (low+high) / 2;
    int guess = *(arr+mid);
    if(guess == elem){
      printf("low: %d, high: %d, mid: %d, guess: %d found it!\n", low, high, mid, guess);
      return guess;
    }
    else if(guess > elem){
      printf("low: %d, high: %d, mid: %d, guess: %d\n", low, high, mid, guess);
      high = mid - 1;
    }
    else{
      printf("low: %d, high: %d, mid: %d, guess: %d\n", low, high, mid, guess);
      low = mid + 1;
    }
  }
  printf("Element %d not found!\n", elem);
  return 1;
}

static void getHalves(int start){
  float sum = start;
  int i = 1;
  printf("0 %d\n", start);
  while(ceil(sum) != 1){
    sum /= 2.0f;
    printf("%d %.0f\n", i, ceil(sum));
    i += 1;
  }
}

int main(int argc, char *argv[])
{
  int array[] = {1,2,3,4,5};
  short arrayLen = sizeof(array) / sizeof(array[0]);
  // for(int i = 0; i < arrayLen; i++){
  //   printf("%p -> %d\n", &array[i], array[i]);
  // }
  //getHalves(100);
  int array2[10] = {1,2,3,4,5,6,7,8,9,10};
  arrayLen = sizeof(array2) / sizeof(array2[0]);
  for(int i = 0; i <= arrayLen; i++){
    int test = binarySearch(array2, arrayLen, i);
  }
  return 0;
}

